mc-sema/cfgToLLVM/linuxArchOps.cpp

```diff
@@ -249,78 +247,82 @@ Value *linuxFreeStack(Module *M, Value *stackAlloc, BasicBlock *&driverBB) {
-            FunctionType::get( Type::getIntNTy(M->getContext(), regWidth), false );
+    // no driver, make one
+    FunctionType    *callbackTy = 
+        FunctionType::get( Type::getIntNTy(M->getContext(), regWidth), false );
 
-        Function *F = dynamic_cast<Function*>(M->getOrInsertFunction(callback_name, callbackTy));
-        TASSERT( F != NULL, "Cannot create callback stub" );
-        F->setLinkage(GlobalValue::InternalLinkage);
-        F->addFnAttr(Attribute::Naked);
+    Function *F = dynamic_cast<Function*>(M->getOrInsertFunction(callback_name, callbackTy));
+    TASSERT( F != NULL, "Cannot create callback stub" );
+    F->setLinkage(GlobalValue::InternalLinkage);
+    F->addFnAttr(Attribute::Naked);
 
-        // add code to driver
-        BasicBlock *block = BasicBlock::Create(
-                F->getContext(), "", F);
+    // add code to driver
+    BasicBlock *block = BasicBlock::Create(
+            F->getContext(), "", F);
 
-        std::vector<Type*>FuncTy_9_args;
-        FuncTy_9_args.push_back(call_tgt->getType());
-        FunctionType* FuncTy_9 = FunctionType::get(
-                /*Result=*/Type::getVoidTy(M->getContext()),
-                /*Params=*/FuncTy_9_args,
-                /*isVarArg=*/false);
+    std::vector<Type*>FuncTy_9_args;
+    FuncTy_9_args.push_back(call_tgt->getType());
+    FunctionType* FuncTy_9 = FunctionType::get(
+            /*Result=*/Type::getVoidTy(M->getContext()),
+            /*Params=*/FuncTy_9_args,
+            /*isVarArg=*/false);
+
+    if(getSystemArch(M) == _X86_64_) {
 
         InlineAsm* ptr_13 = InlineAsm::get(FuncTy_9, "movq $0, %rax\n", "imr,~{dirflag},~{fpsr},~{flags}",true);
         CallInst* void_12 = CallInst::Create(ptr_13, call_tgt, "", block);
         void_12->setCallingConv(CallingConv::C);
         void_12->setTailCall(false);
+    } else {
+        InlineAsm* ptr_13 = InlineAsm::get(FuncTy_9, "pushl $0\n", "imr,~{dirflag},~{fpsr},~{flags}",true);
+        CallInst* void_12 = CallInst::Create(ptr_13, call_tgt, "", block);
+        void_12->setCallingConv(CallingConv::C);
+        void_12->setTailCall(false);
+    }
 
+    Function *inception = linuxGetInceptionFunction(M);
 
-        Function *inception = linuxGetInceptionFunction(M);
+    std::vector<Type*>call_inception_args;
+    call_inception_args.push_back(inception->getType());
+    FunctionType* call_inception_ty = FunctionType::get(
+            /*Result=*/Type::getVoidTy(M->getContext()),
+            /*Params=*/call_inception_args,
+            /*isVarArg=*/false);
 
-        std::vector<Type*>call_inception_args;
-        call_inception_args.push_back(inception->getType());
-        FunctionType* call_inception_ty = FunctionType::get(
-                /*Result=*/Type::getVoidTy(M->getContext()),
-                /*Params=*/call_inception_args,
-                /*isVarArg=*/false);
+    if(getSystemArch(M) == _X86_64_) {
 
         InlineAsm* ptr_15 = InlineAsm::get(call_inception_ty, 
```

Comment:
can you combine all the 64-bit stuff into an `if`, then all the 32-bit stuff into an `else` instead of having two `if`/`else` blocks. It's make understanding the sequence of code easier.